lib/portlayer/storage/nfs/volume.go

```diff
@@ -0,0 +1,260 @@
+	// XXX: Potential credential leak, maybe log less...
+	op.Infof("Creating datastore (%s) at target (%q)", storeName, nfsTargetURL.String())
+
+	target, err := mount.Mount(nfsTargetURL)
+	if err != nil {
+		return nil, err
+	}
+
+	defer mount.Unmount(target)
+
+	//we assume that nfsTargetURL.path already exists.
+	//make volumes directory
+	if _, err := target.MkDir(path.Join(nfsTargetURL.Path, VolumesDir), filePermissions); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	//make metadata directory
+	if _, err := target.MkDir(path.Join(nfsTargetURL.Path, metadataDir), filePermissions); err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	selfLink, err := util.VolumeStoreNameToURL(storeName)
+	if err != nil {
+		return nil, err
+	}
+
+	v := &NFSv3VolumeStore{
+		Name:           storeName,
+		volumeStoreDir: nfsTargetURL.Path,
+		MH:             mount,
+		Target:         nfsTargetURL,
+		SelfLink:       selfLink,
+	}
+
+	return v, nil
+}
+
+// Returns the path to the vol relative to the given store.  The dir structure
+// for a vol in a nfs store is `<configured nfs server path>/volumes/<vol ID>/<volume contents>`.
+func (v *NFSv3VolumeStore) volDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, VolumesDir, ID)
+}
+
+func (v *NFSv3VolumeStore) volumesDir() string {
+	return path.Join(v.volumeStoreDir, VolumesDir)
+}
+
+// Returns the path to the metadata directory for a volume
+func (v *NFSv3VolumeStore) volMetadataDirPath(ID string) string {
+	return path.Join(v.volumeStoreDir, metadataDir, ID)
+}
+
+// Creates a volume directory and volume object for NFS based volumes
+func (v *NFSv3VolumeStore) VolumeCreate(op trace.Operation, ID string, store *url.URL, capacityKB uint64, info map[string][]byte) (*storage.Volume, error) {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	defer v.MH.Unmount(target)
+
+	if _, err := target.MkDir(v.volDirPath(ID), filePermissions); err != nil {
+		return nil, err
+	}
+
+	backing := NewNFSVolumeDevice(v.Target, v.volDirPath(ID))
+
+	vol, err := storage.NewVolume(v.SelfLink, ID, info, backing)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := writeMetadata(op, v.volMetadataDirPath(ID), info, target); err != nil {
+		return nil, err
+	}
+
+	op.Infof("nfs volume (%s) successfully created on volume store (%s)", ID, v.Name)
+	return vol, nil
+}
+
+// Removes a volume and all of it's contents from the nfs store. We already know via the cache if it is in use.
+func (v *NFSv3VolumeStore) VolumeDestroy(op trace.Operation, vol *storage.Volume) error {
+	target, err := v.MH.Mount(v.Target)
+	if err != nil {
+		return err
+	}
+
+	defer v.MH.Unmount(target)
+
+	op.Infof("Attempting to remove volume (%s) and it's metadata from volume store (%s)", vol.ID, v.Name)
+
+	//remove volume directory and children
+	if err := target.RemoveAll(v.volDirPath(vol.ID)); err != nil {
```

Comment:
This is an interesting question, It will likely error with the dir does not exist, or if the path does not exist in particular since this can return a path error in the standard `os` package in golang as well(maybe not ours). However, If the dir is not there and the intention was to remove do we want to return an error or not if this scenario occurs? @fdawg4l my thoughts are that we are attempting to remove something that the cache thought was there, in which case we may want to error since something bad has occurred(most likely the directory was removed OOB via vsphere UI, govc, etc...).